notanorm/jsondb.py

```diff
@@ -0,0 +1,571 @@
+            val = self.__val_from(set_exp.right, parameters)
+            sets[col_name] = val
+        where = op.find(exp.Where)
+        where_dict = self.__op_where(where, list(parameters))
+        for row in self.__dat.setdefault(tab.name, []):
+            if all(self.__op(row[k], v) for k, v in where_dict.items()):
+                row.update(sets)
+                ret.rowcount += 1
+        return ret
+
+    def __op_delete(self, ret, op, parameters):
+        tab = op.find(exp.Table)
+        where = op.find(exp.Where)
+        where_dict = self.__op_where(where, list(parameters))
+        new = []
+        for row in self.__get_tab_dat(tab.name):
+            if not all(v == row[k] for k, v in where_dict.items()):
+                new.append(row)
+                ret.rowcount += 1
+        self.__dat[tab.name] = new
+
+    def __op_where(self, op, parameters: list, *, ret=None) -> dict:
+        ret = {} if ret is None else ret
+        if op:
+            op = op.this
+            if isinstance(op, exp.And):
+                self.__op_where(op.left, parameters, ret=ret)
+            elif isinstance(op, exp.Is):
+                assert isinstance(op.right, exp.Null)
+                ret[op.left.name] = None
+            elif isinstance(op, exp.EQ):
+                val = self.__val_from(op.right, parameters)
+                if isinstance(op.left, exp.Literal):  # pragma: no cover
+                    raise RuntimeError("literal comparisons not supported yet")
+                ret[op.left.name] = val
+            elif isinstance(op, (exp.GT, exp.GTE, exp.LT, exp.LTE)):
+                val = self.__val_from(op.right, parameters)
+                cod = {exp.GT: ">", exp.GTE: ">=", exp.LT: "<", exp.LTE: "<="}.get(
+                    type(op)
+                )
+                val = Op(cod, val)
+                if isinstance(op.left, exp.Literal):  # pragma: no cover
+                    raise RuntimeError("literal comparisons not supported yet")
+                ret[op.left.name] = val
+        return ret
+
+    @staticmethod
+    def __val_from(op, parameters):
+        if isinstance(op, exp.Placeholder):
+            val = parameters.pop(0)
+        elif isinstance(op, exp.Boolean):
+            val = op.args["this"]
+        else:
+            val = op.output_name
+            if op.is_int:
+                val = int(val)
+            elif op.is_number:
+                val = float(val)
+        return val
+
+    def __op_select(self, op, parameters):
+        cols = []
+        params = list(parameters)
+        aliases = {}
+        for col in op.expressions:
+            if isinstance(col, exp.Alias):
+                aliases[col.this.name] = col.output_name
+                col = col.this
+            cols.append(col.name)
+        tab = op.args["from"].args["expressions"][0].name
+        where = op.find(exp.Where)
+        where_dict = self.__op_where(where, params)
+        rows = self.__get_tab_dat(tab)
```

Comment:
Assuming we'll switch to short-lived connections to avoid stale reads?

Reply:
FWIW not sure this is strictly true anymore -- don't we do like a refcounting thing with `ShareLocationDbGuard` now?

Reply:
yes, ShareLocationDbGuard is responsible for calling "refresh" to reread from disk when appropriate.   And to call "flush/commit" when closing.

Reply:
added "flush" command so the caller doesn't have to "commit" (close a current transaction) when "flushing". will change cvfs to use this too.   can see it being an issue in certain multithread circumstances